metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -893,7 +896,53 @@ std::vector<TAlphabet> BOSS::get_node_seq(edge_index x) const {
         ret[--i] = get_node_last_value(x);
     }
 
-    return ret;
+    return std::make_tuple(std::move(ret), x, false);
+}
+
+/**
+ * Given an edge index i, this function returns the k-mer sequence of its
+ * source node.
+ */
+std::vector<TAlphabet> BOSS::get_node_seq(edge_index i) const {
+    return std::get<0>(get_node_seq_with_end_node_indexed(i));
+}
+
+/**
+ * Given an edge index i, this function returns the k-mer sequence of its
+ * source node, and the node whose last character corresponds to the first
+ * character of the sequence. If force_get_end_node is false, then the last
+ * traversed node will only be returned if it is computed.
+ */
+std::pair<std::vector<TAlphabet>, std::optional<edge_index>> BOSS
+::get_node_seq_with_end_node(edge_index i, bool force_get_end_node) const {
+    auto [seq, last_node, partially_traversed] = get_node_seq_with_end_node_indexed(i);
+
+    std::optional<edge_index> last_node_opt = std::nullopt;
+
+    if (!partially_traversed) {
+        // if the full traversal was done, then we already have the last node
+        last_node_opt = last_node;
+
+    } else if (force_get_end_node) {
+        // if the full traversal was not done due to the suffix range index, but
+        // the last node from traversal is still requested, then complete the traversal
+        // to get that node
+        assert(indexed_suffix_length_);
+
+        auto it = seq.rend() - indexed_suffix_length_;
+        assert(*it == get_node_last_value(last_node));
+        for (++it; it != seq.rend(); ++it) {
+            last_node = bwd(last_node);
+            assert(*it == get_node_last_value(last_node));
+        }
+
+        last_node_opt = last_node;
+    }
```

Comment:
there are too many optional and stuff, which complicates the whole design of these interfaces.

To simplify this, start with removing this optional argument -- just have two functions.

Reply:
So, get back the old method and add the one you need as a new separate method.

Reply:
I still want to keep `get_node_seq_with_end_node_indexed` to avoid having two functions (this and `get_node_seq`) that do nearly the same work. Later down the road, someone will make some changes to `get_node_seq` and it'll be their responsibility to make sure that the same changes get propagated to the other function (which most likely won't happen).

Reply:
You also have code duplications right now. E.g., this block:
```
        auto it = seq.rend() - indexed_suffix_length_;
        assert(*it == get_node_last_value(last_node));
        for (++it; it != seq.rend(); ++it) {
            last_node = bwd(last_node);
            assert(*it == get_node_last_value(last_node));
        }
```
But on top you also have a bunch of confusing arguments that are most of the time aren't used.

Reply:
also lots of confusing if - else and stuff. So it's really better to get rid of all those optional arguments like `bool force_get_end_node`

Reply:
I simplified this quite a bit. Now, there's a method called `get_node_seq_with_last_traversed_node` which returns the sequence, the last traversed node, and the number of steps still needed to reach k bwd steps. This way, if any downstream method can then call `get_minus_k_value(last_traversed_node, steps_remaining)` to get what it wants.